src/DataProtection/DataProtection/src/KeyManagement/KeyRingProvider.cs

```diff
@@ -292,6 +311,146 @@ internal IKeyRing GetCurrentKeyRingCore(DateTime utcNow, bool forceRefresh = fal
+            }
+        }
+
+        if (existingCacheableKeyRing is not null)
+        {
+            Debug.Assert(!forceRefresh, "Read cached key ring even though forceRefresh is true");
+            Debug.Assert(!CacheableKeyRing.IsValid(existingCacheableKeyRing, utcNow), "Should already have returned a valid cached key ring");
+            return existingCacheableKeyRing.KeyRing;
+        }
+
+        // Since there's no cached key ring we can return, we have to wait.  It's not ideal to wait for a task we
+        // just scheduled, but it makes the code a lot simpler (compared to having a separate, synchronous code path).
+        // Cleverness: swallow any exceptions - they'll be surfaced by GetKeyRingFromCompletedTask, if appropriate.
+        existingTask
+            .ContinueWith(static _ => { }, TaskScheduler.Default)
+            .Wait();
+
+        var newKeyRing = GetKeyRingFromCompletedTask(existingTask, utcNow); // Throws if the task failed
+        if (newKeyRing is null)
+        {
+            // Another thread won - check whether it cached a new key ring
+            var newCacheableKeyRing = Volatile.Read(ref _cacheableKeyRing);
+            if (newCacheableKeyRing is null)
+            {
+                // There will have been a better exception from the winning thread
+                throw Error.KeyRingProvider_RefreshFailedOnOtherThread();
+            }
+
+            newKeyRing = newCacheableKeyRing.KeyRing;
+        }
+
+        return newKeyRing;
+    }
+
+    /// <summary>
+    /// Returns null if another thread already processed the completed task.
+    /// Otherwise, if the given completed task completed successfully, clears the task
+    /// and either caches and returns the resulting key ring or throws, according to the
+    /// successfulness of the task.
+    /// </summary>
+    private IKeyRing? GetKeyRingFromCompletedTask(Task<CacheableKeyRing> task, DateTime utcNow)
+    {
+        Debug.Assert(task.IsCompleted);
+
+        lock (_cacheableKeyRingLockObj)
+        {
+            // If the parameter doesn't match the field, another thread has already consumed the task (and it's reflected in _cacheableKeyRing)
+            if (!ReferenceEquals(task, Volatile.Read(ref _cacheableKeyRingTask)))
+            {
+                return null;
+            }
+
+            Volatile.Write(ref _cacheableKeyRingTask, null);
+
+            if (task.Status == TaskStatus.RanToCompletion)
+            {
+                var newCacheableKeyRing = task.Result;
+                Volatile.Write(ref _cacheableKeyRing, newCacheableKeyRing);
+
+                // An unconsumed task result is considered to satisfy forceRefresh.  One could quibble that this isn't really
+                // a forced refresh, but we'll still return a key ring newer than the one the caller was dissatisfied with.
+                return newCacheableKeyRing.KeyRing;
+            }
+
+            var exception = task.Exception!;
```

Comment:
Is `Task.Exception` guaranteed to be not null here? If the task was canceled, I believe this would be null. Should we guard this codepath against that?

Reply:
I thought there'd be a cancellation exception.  Is that not the case?  It's a little academic, since we're not actually using a CancellationToken.

Reply:
```
var cts =  new CancellationTokenSource();
var task = new Task(() => Thread.Sleep(1000), cts.Token);
cts.Cancel();
Console.WriteLine(task.IsCompletedSuccessfully);
Console.WriteLine(task.IsCanceled);
Console.WriteLine(task.Exception == null);
```

produces:

```
False
True
True
```

So it looks like the CancellationException isn't set on the task.

> It's a little academic, since we're not actually using a CancellationToken.

Fair enough. Perhaps it still helps to be explicit about querying on a faulted exception here.

Reply:
Huh
>If you're waiting on a Task that transitions to the Canceled state, a System.Threading.Tasks.TaskCanceledException exception (wrapped in an AggregateException exception) is thrown. This exception indicates successful cancellation instead of a faulty situation. Therefore, the task's Exception property returns null.

Reply:
I see three ways forward:
1. Add new log messages about cancellation
2. Use the current log messages and instantiate our own OperationCancelledException - the stack will be wrong
3. Just assert that the exception is non-null since cancellation is impossible

I'm leaning towards (3) - thoughts, @captainsafia?

Reply:
3rd options works for me given that it reflects the constraints of the system.

Reply:
Task.GetAwaiter().GetResult() can be used to both get the result in the success case and throw an exception in the non-success case (both canceled and other exceptions). You'd need to add try catch of course.

Reply:
Done.  I've also added a note in case someone tries to make it cancellable in the future.

Reply:
> Task.GetAwaiter().GetResult() can be used to both...

I thought it was silly to catch and rethrow, but I can change it if you think that's much clearer or more conventional (or correct, though I'm missing how).

Reply:
Stephen says it's more conventional.  Updated.